### PR TITLE
[SYCL][libdevice] Get rid of builtins for memcpy memset in libdevice

### PR DESCRIPTION
**Pull request by jinge90**
Previously, memcpy and memset are implemented in libdevice by calling
to __builtin_memcpy and __builtin_memset. Such code will lead to
infinite loop if underying CPU runtime implements memcpy/set builtin by
calling memcpy, memset. The function call chain is following:
In libdevice:
memcpy(dest, src, n) {
  return __devicelib_memcpy(dest, src, n);
}

__devicelib_memcpy(dest, src, n) {
  return __builtin_memcpy(dest, src, n);
}

In CPU runtime:
Handing of __builtin_memcpy<-------------------|
                  |				                         |
		  |				                         |
		  |---->memcpy------>__devicelib_memcpy
In order to fix this, we have to provide implementation for memcpy/set
without using __builtin_*.

### CODE REVIEW

**Comment by jinge90** [comment]
The patch is done but we need more extensive testing for it, we are working on enhance tests for libdevice memcpy/set using USM: https://github.com/intel/llvm-test-suite/pull/558

Thanks very much.

**Comment by jinge90** [comment]
Hi, @vzakhari 
Could you help review this patch?
Thanks very much.

File: libdevice/fallback-cstring.cpp [hunk 1]
```diff
@@ -10,51 +10,171 @@
...
+static void *__devicelib_memcpy_uint8_aligned(void *dest, const void *src,
+                                              size_t n) {
+  if (dest == NULL || src == NULL || n == 0)
+    return dest;
+
+  uint8_t *dest_uint8 = reinterpret_cast<uint8_t *>(dest);
+  const uint8_t *src_uint8 = reinterpret_cast<const uint8_t *>(src);
+  for (size_t idx = 0; idx < n; ++idx) {
+    dest_uint8[idx] = src_uint8[idx];
+  }
```
Comment (bader) [inline libdevice/fallback-cstring.cpp:23]:
```suggestion
  for (size_t idx = 0; idx < n; ++idx)
    dest_uint8[idx] = src_uint8[idx];
```

File: libdevice/fallback-cstring.cpp [hunk 2]
```diff
@@ -10,51 +10,171 @@
...
+    return dest;
+
+  uint32_t *dest_addr = reinterpret_cast<uint32_t *>(dest);
+  const uint32_t *src_addr = reinterpret_cast<const uint32_t *>(src);
+  size_t tailing_bytes = n % sizeof(uint32_t);
+  size_t copy_num = n >> 2;
+  size_t idx;
+  for (idx = 0; idx < copy_num; ++idx)
+    dest_addr[idx] = src_addr[idx];
+  __devicelib_memcpy_uint8_aligned(&dest_addr[idx], &src_addr[idx],
```
Comment (bader) [inline libdevice/fallback-cstring.cpp:40]:
```suggestion
    dest_addr[idx] = src_addr[idx];
    
  __devicelib_memcpy_uint8_aligned(&dest_addr[idx], &src_addr[idx],
```

File: libdevice/fallback-cstring.cpp [hunk 3]
```diff
@@ -10,51 +10,171 @@
...
+  if (head_ua_len >= n)
+    return __devicelib_memcpy_uint8_aligned(dest, src, n);
+  else {
+    __devicelib_memcpy_uint8_aligned(dest, src, head_ua_len);
+    void *dest_aligned_addr = reinterpret_cast<void *>(dest_addr + head_ua_len);
+    const void *src_aligned_addr =
+        reinterpret_cast<const void *>(src_addr + head_ua_len);
+    n -= head_ua_len;
+    __devicelib_memcpy_uint32_aligned(dest_aligned_addr, src_aligned_addr, n);
+  }
```
Comment (bader) [inline libdevice/fallback-cstring.cpp:71]:
```suggestion
  __devicelib_memcpy_uint8_aligned(dest, src, head_ua_len);
  void *dest_aligned_addr = reinterpret_cast<void *>(dest_addr + head_ua_len);
  const void *src_aligned_addr =
      reinterpret_cast<const void *>(src_addr + head_ua_len);
  n -= head_ua_len;
  __devicelib_memcpy_uint32_aligned(dest_aligned_addr, src_aligned_addr, n);
```

File: libdevice/fallback-cstring.cpp [hunk 4]
```diff
@@ -10,51 +10,171 @@
...
+}
+
+static void *__devicelib_memset_uint8_aligned(void *dest, int c, size_t n) {
+  if (dest == NULL || n == 0)
+    return dest;
+
+  uint8_t *dest_addr = reinterpret_cast<uint8_t *>(dest);
+  for (size_t idx = 0; idx < n; ++idx) {
+    dest_addr[idx] = static_cast<uint8_t>(c);
+  }
```
Comment (bader) [inline libdevice/fallback-cstring.cpp:83]:
```suggestion
  for (size_t idx = 0; idx < n; ++idx)
    dest_addr[idx] = static_cast<uint8_t>(c);
```

File: libdevice/fallback-cstring.cpp [hunk 5]
```diff
@@ -10,51 +10,171 @@
...
-      n--;
-    } else {
-      return *s1_uint8_ptr - *s2_uint8_ptr;
-    }
+
+  for (size_t idx = 0; idx < n; ++idx) {
+    if (s1_uint8_ptr[idx] == s2_uint8_ptr[idx])
+      continue;
+    else
+      return s1_uint8_ptr[idx] - s2_uint8_ptr[idx];
```
Comment (bader) [inline libdevice/fallback-cstring.cpp:144]:
```suggestion
    if (s1_uint8_ptr[idx] != s2_uint8_ptr[idx])
      return s1_uint8_ptr[idx] - s2_uint8_ptr[idx];
```

File: libdevice/fallback-cstring.cpp [hunk 6]
```diff
@@ -10,51 +10,171 @@
...
-                                                     s2_uint32_ptr, n);
+  if (idx < cmp_num)
+    return __devicelib_memcmp_uint8_aligned(
+        &s1_uint32_ptr[idx], &s2_uint32_ptr[idx], sizeof(uint32_t));
+
+  if (tailing_bytes == 0)
+    return 0;
+  else
+    return __devicelib_memcmp_uint8_aligned(
+        &s1_uint32_ptr[cmp_num], &s2_uint32_ptr[cmp_num], tailing_bytes);
```
Comment (bader) [inline libdevice/fallback-cstring.cpp:177]:
```suggestion

  return __devicelib_memcmp_uint8_aligned(
      &s1_uint32_ptr[cmp_num], &s2_uint32_ptr[cmp_num], tailing_bytes);
```

**Review by bader — COMMENTED** [review]
A few style changes required to align with the project coding guidelines.

File: libdevice/fallback-cstring.cpp (hunk 1 above)
Comment (jinge90) [reply libdevice/fallback-cstring.cpp]: Done.

File: libdevice/fallback-cstring.cpp (hunk 2 above)
Comment (jinge90) [reply libdevice/fallback-cstring.cpp]: Done.

File: libdevice/fallback-cstring.cpp (hunk 3 above)
Comment (jinge90) [reply libdevice/fallback-cstring.cpp]: Done.

File: libdevice/fallback-cstring.cpp (hunk 5 above)
Comment (jinge90) [reply libdevice/fallback-cstring.cpp]: Done.

File: libdevice/fallback-cstring.cpp (hunk 6 above)
Comment (jinge90) [reply libdevice/fallback-cstring.cpp]: Done.

File: libdevice/fallback-cstring.cpp (hunk 4 above)
Comment (jinge90) [reply libdevice/fallback-cstring.cpp]: Done.

**Review by bader — APPROVED** [review]

**Comment by jinge90** [comment]
Hi, @vzakhari 
Could you take a look?
Thanks very much.

**Review by vzakhari — APPROVED** [review]

**Pull request merged**